### ISSUE TITLE
"sorted“ attributes should not be kept when using fsubset

### ISSUE THREAD

**Issue by hongyuanjia** [SebKrantz/collapse#145]
Maybe related to #136.

data.table internally uses attributes to store some metadata. `sorted` is one of them indicating names of the columns that have been sorted using `data.table::setkey()`.

When subsetting using `[.data.table`, the `sorted` attribute will be removed. But using `collapse::fsubset.data.frame()` did not. I am not sure if this will introduce errors when doing other data.table operations, since giving #136 that using `fsubset` will break the 
self-referencing of data.table.

``` r
dt <- data.table::as.data.table(mtcars)
data.table::setkey(dt, cyl)

str(dt[c(2L, 1L)])
#> Classes 'data.table' and 'data.frame':   2 obs. of  11 variables:
#>  $ mpg : num  24.4 22.8
#>  $ cyl : num  4 4
#>  $ disp: num  147 108
#>  $ hp  : num  62 93
#>  $ drat: num  3.69 3.85
#>  $ wt  : num  3.19 2.32
#>  $ qsec: num  20 18.6
#>  $ vs  : num  1 1
#>  $ am  : num  0 1
#>  $ gear: num  4 4
#>  $ carb: num  2 1
#>  - attr(*, ".internal.selfref")=<externalptr>

str(collapse::fsubset.data.frame(dt, c(2L, 1L)))
#> Classes 'data.table' and 'data.frame':   2 obs. of  11 variables:
#>  $ mpg : num  24.4 22.8
#>  $ cyl : num  4 4
#>  $ disp: num  147 108
#>  $ hp  : num  62 93
#>  $ drat: num  3.69 3.85
#>  $ wt  : num  3.19 2.32
#>  $ qsec: num  20 18.6
#>  $ vs  : num  1 1
#>  $ am  : num  0 1
#>  $ gear: num  4 4
#>  $ carb: num  2 1
#>  - attr(*, ".internal.selfref")=<externalptr> 
#>  - attr(*, "sorted")= chr "cyl"
```

<sup>Created on 2021-03-29 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by SebKrantz** [comment]
Thanks, I'll see what I can do about this.

**Comment by hongyuanjia** [comment]
Below are code that the wrongly attached `sorted` attribute becomes an issue:

``` r
dt <- data.table::as.data.table(mtcars)
data.table::setkey(dt, cyl)
dt1 <- dt[c(12L, 11L)]
dt2 <- collapse::fsubset.data.frame(dt, c(12L, 11L))

data.table::haskey(dt1)
#> [1] FALSE
data.table::haskey(dt2)
#> [1] TRUE
data.table::key(dt1)
#> NULL
data.table::key(dt2)
#> [1] "cyl"

data.table::setkey(data.table::copy(dt1), "cyl")[]
#>     mpg cyl disp  hp drat   wt  qsec vs am gear carb
#> 1: 21.4   4  121 109 4.11 2.78 18.60  1  1    4    2
#> 2: 21.0   6  160 110 3.90 2.62 16.46  0  1    4    4
data.table::setkey(data.table::copy(dt2), "cyl")[]
#>     mpg cyl disp  hp drat   wt  qsec vs am gear carb
#> 1: 21.0   6  160 110 3.90 2.62 16.46  0  1    4    4
#> 2: 21.4   4  121 109 4.11 2.78 18.60  1  1    4    2

dt1[.(4)]
#> Error in `[.data.table`(dt1, .(4)): When i is a data.table (or character vector), the columns to join by must be specified using 'on=' argument (see ?data.table), by keying x (i.e. sorted, and, marked as sorted, see ?setkey), or by sharing column names between x and i (i.e., a natural join). Keyed joins might have further speed benefits on very large data due to x being sorted in RAM.
dt2[.(4)]
#>    mpg cyl disp hp drat wt qsec vs am gear carb
#> 1:  NA   4   NA NA   NA NA   NA NA NA   NA   NA

dt3 <- data.table::setkey(data.table::copy(dt1), "cyl")
dt3[.(4)]
#>     mpg cyl disp  hp drat   wt qsec vs am gear carb
#> 1: 21.4   4  121 109 4.11 2.78 18.6  1  1    4    2
```

<sup>Created on 2021-03-29 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by SebKrantz** [comment]
This is also fixed in 1.6.0.
